File: openHAB/ImageUITableViewCell.m
//
//  ImageUITableViewCell.m
//  openHAB
//
//  Created by Victor Belov on 18/04/14.
//  Copyright (c) 2014 Victor Belov. All rights reserved.
//

#import "ImageUITableViewCell.h"
#import <SDWebImage/UIImageView+WebCache.h>
#import <SDWebImage/SDWebImageDownloader.h>

@implementation ImageUITableViewCell
@synthesize widgetImage, delegate;

NSTimer *refreshTimer;

- (void)loadWidget:(OpenHABWidget *)widgetToLoad
{
    self.widget = widgetToLoad;
    // Remove image from SDImage cache
}

- (void)displayWidget
{
    widgetImage = (UIImageView*)[self viewWithTag:901];
    if (widget.image == nil) {
        [self loadImage];
    } else {
        [widgetImage setImage:widget.image];
    }
    // If widget have a refresh rate configured, schedule an image update timer
    if (self.widget.refresh != nil && refreshTimer == nil) {
        NSTimeInterval refreshInterval = [self.widget.refresh floatValue] / 1000;
        refreshTimer = [NSTimer scheduledTimerWithTimeInterval:refreshInterval
                                                        target:self
                                                      selector:@selector(refreshImage:)
                                                      userInfo:nil
                                                       repeats:YES];
    }
}

- (void)loadImage
{
    int random = arc4random() % 1000;
    [widgetImage setImageWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"%@&random=%d", self.widget.url, random]] placeholderImage:nil options:SDWebImageCacheMemoryOnly completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType) {
        widget.image = self.widgetImage.image;
        [widgetImage setFrame:self.contentView.frame];
        if (self.delegate != nil) {
            [self.delegate didLoadImage];
        }
    }];
}

- (void)refreshImage:(NSTimer *)timer
{
    int random = arc4random() % 1000;
    [widgetImage setImageWithURL:[NSURL URLWithString:[NSString stringWithFormat:@"%@&random=%d", self.widget.url, random]] placeholderImage:self.widgetImage.image options:SDWebImageCacheMemoryOnly completed:^(UIImage *image, NSError *error, SDImageCacheType cacheType) {
        widget.image = self.widgetImage.image;
    }];
}

- (void)willMoveToWindow:(UIWindow *)newWindow
{
    [super willMoveToWindow:newWindow];
    if (newWindow == nil && refreshTimer != nil) {
        [refreshTimer invalidate];
        refreshTimer = nil;
    }
}

@end

File: openHAB/ImageUITableViewCell.swift
//
//  ImageUITableViewCell.swift
//  openHAB
//
//  Created by Victor Belov on 18/04/14.
//  Copyright (c) 2014 Victor Belov. All rights reserved.
//
//  Converted to Swift 4 by Tim Müller-Seydlitz and Swiftify on 06/01/18
//

import SDWebImage

protocol ImageUITableViewCellDelegate: NSObjectProtocol {
    func didLoadImageOf(_ cell: ImageUITableViewCell?)
}

var refreshTimer: Timer?

class ImageUITableViewCell: GenericUITableViewCell {
    @IBOutlet weak var widgetImage: UIImageView!
    weak var delegate: ImageUITableViewCellDelegate?

    @objc override var widget: OpenHABWidget! {
        get {
            return super.widget
        }
        set(widget) {
            super.widget = widget
            // Remove image from SDImage cache
        }
    }

    override func displayWidget() {
        if widget?.image == nil {
            loadImage()
        } else {
            widgetImage?.image = widget?.image
        }
        // If widget have a refresh rate configured, schedule an image update timer
//        if let refresh = widget.refresh, let refreshTimer = refreshTimer {
//            let refreshInterval = TimeInterval(refresh / 1000)
//            self.refreshTimer = Timer.scheduledTimer(timeInterval: refreshInterval, target: self,
//                                                selector: #selector(ImageUITableViewCell.refreshImage(_:)), userInfo: nil, repeats: true)
//        }
    }

    func imageURL() -> URL {
        let random = Int.random(in: 0..<1000)
        var components = URLComponents(string: widget.url)
        components?.queryItems?.append(contentsOf: [
            URLQueryItem(name: "random", value: String(random))
            ])
        return components?.url ?? URL(string: "")!
    }

    func loadImage() {
        widgetImage?.sd_setImage(with: imageURL(), placeholderImage: nil) { (image, error, cacheType, imageURL) in
                    // Perform operation.
                    self.widget.image = image
                }
    }

    @objc func refreshImage(_ timer: Timer?) {
        widgetImage?.sd_setImage(with: imageURL(), placeholderImage: widgetImage?.image)
    }

    func willMove(to newWindow: UIWindow?) {
        super.willMove(toWindow: newWindow)
        if newWindow == nil && refreshTimer != nil {
            refreshTimer?.invalidate()
            refreshTimer = nil
        }
    }
}
